scipy/optimize/_sda.py

```diff
@@ -0,0 +1,572 @@
+
+                # No improvement since long time
+                if self.not_improved_idx >= self.not_improved_max_idx:
+                    if j == 0 or self.state.current_energy < self.emin:
+                        self.emin = self.state.current_energy
+                        self.xmin = np.copy(self.state.current_location)
+        # End of MarkovChain loop
+
+    def local_search(self):
+        # Decision making for performing a local search
+        # based on Markov chain results
+        # If energy has been improved or no improvement since too long,
+        # performing a local search with the best Markov chain location
+        if self.state_improved:
+            # Global energy has improved, let's see if LS improved further
+            e, x = self.ofw.local_search(self.state.xbest)
+            if e < self.state.ebest:
+                self.not_improved_idx = 0
+                self.state.ebest = e
+                self.state.xbest = np.copy(x)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+                return
+        # Check probability of a need to perform a LS even if no improvment
+        # (Dual annealing principle)
+        do_ls = False
+        if self.K < 90 * len(self.state.current_location):
+            pls = np.exp(self.K * (self.state.ebest - self.state.current_energy
+                                   ) / self.temperature_step)
+            if pls >= self._rs.random_sample():
+                do_ls = True
+        # Global energy not improved, let's see what LS gives
+        # on the best Markov chain location
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            do_ls = True
+        if do_ls:
+            e, x = self.ofw.local_search(self.xmin)
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search and default gradient computation.
+    Default local minimizer is L-BFGS-B
+    """
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = None
+        lu = list(zip(*bounds))
+        self.lower = np.array(lu[0])
+        self.upper = np.array(lu[1])
+        self.ls_max_iter = self.lower.size * 6
+
+        if self.ls_max_iter < 100:
+            self.ls_max_iter = 100
+        if self.ls_max_iter > 1000:
+            self.ls_max_iter = 1000
```

Comment:
I think you should rewrite this as:
```
    MAXITER_RATIO = 6
    MAXITER_MAX = 1000
    MAXITER_MIN = 100
    ...
    n = self.lower.size
    self.ls_max_iter = min(max(n*MAXITER_RATIO, MAXITER_MIN), MAXITER_MAX)

Reply:
Actually I would recommend you to try to use constants to all numbers that are arbitrary in your implementation. You have done it to some extension, but there are still many others numbers that could be replaced by constants.

This improve the readability of your code, since it helps differentiating between arbitrary constants and factors that have some strong mathematical reason to be there.

Reply:
Also, those values seem fairly arbitrary and potentially very dependent on the objective function.   When I first looked at this I couldn't understand why you were hardcoding these value in, but then I realized it's only used for the default minimizer if nothing else is specified.  How about you move this into another method called, e.g. setup_default_minimizer.  That will make it clear that they are only used for that purpose.

Reply:
@js850 by my personal experience, it may be too hard to explain in the documentation  and allow the user to specify all internal constants.

For instance, for the ``trust-constr`` (in preparation #7729) and ``trust-exact`` (#6919) I had a lot of internal constants, and many of those the values  were described in the original papers and worked well for me in the tests I did. In this case, I decided to keep those hard-coded (as constants), allowing the user to pass only the most important values. This way the user would have only something about 5 to 10 parameters to tune.

So I believe the best approach to be follow here is:
- Allow the user to specify parameters that have a clear meaning and large impact in the result.
- Hard code the ones that either: 1. does not make much difference in the final value; or, 2. have a good default and a not very clear meaning.